File: FastMines_core/src/fmg/core/img/PolarLightLogoTransformer.cs
using System;
using Fmg.Common;

namespace Fmg.Core.Img {

    /// <summary> Transforming of logo palette </summary>
    public class PolarLightLogoTransformer : IModelTransformer {

        public void Execute(IAnimatedModel model) {
            if (!(model is LogoModel lm))
                throw new InvalidOperationException("Illegal usage transformer");

            if (!lm.PolarLights)
                return;

            double rotateAngleDelta = 360.0 / lm.TotalFrames; // 360° / TotalFrames
            if (!lm.AnimeDirection)
                rotateAngleDelta = -rotateAngleDelta;
            HSV[] palette = lm.Palette;
            for (int i = 0; i < palette.Length; ++i)
                palette[i].h += rotateAngleDelta;
        }

    }

}

File: FastMines_core/src/fmg/core/img/PolarLightLogoTransformer.js
/** Transforming of logo palette */
class PolarLightLogoTransformer extends IModelTransformer {

    static get className() { return 'PolarLightLogoTransformer'; }
    get className() { return PolarLightLogoTransformer.className; }

    execute(/*IImageModel*/ model) {
        var lm = model;

        if (!lm.polarLights)
            return;

        var rotateAngleDelta = 360.0 / lm.totalFrames; // 360° / TotalFrames
        if (!lm.animeDirection)
            rotateAngleDelta = -rotateAngleDelta;
        var palette = lm.palette;
        for (var i = 0; i < palette.length; ++i)
            palette[i].h += rotateAngleDelta;
    }

}
